### PR TITLE
Pallet Preimage

### PR DESCRIPTION
**Pull request by shawntabrizi**
Superseded in #10356.

### CODE REVIEW

File: frame/preimage/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+	fn preimage_requested(hash: T::Hash) -> bool {
+		Requests::<T>::contains_key(hash)
+	}
+
+	fn get_preimage(hash: T::Hash) -> Option<Vec<u8>> {
+		Preimages::<T>::get(hash).map(|preimage| preimage.preimage.to_vec())
+	}
+
+	fn note_preimage(bytes: Vec<u8>) -> Result<(), ()> {
```
Comment (xlc) [inline frame/preimage/src/lib.rs:281]: any reason why you want to remove the error?

File: frame/support/src/traits/misc.rs [hunk 2]
```diff
@@ -564,6 +564,26 @@ impl<T: TypeInfo + 'static> TypeInfo for WrapperKeepOpaque<T> {
...
+pub trait PreimageHandler<Hash> {
+	/// Returns whether a preimage exists for a given hash.
+	fn preimage_exists(hash: Hash) -> bool;
+	/// Returns whether a preimage request exists for a given hash.
+	fn preimage_requested(hash: Hash) -> bool;
+	/// Returns the preimage for a given hash.
+	fn get_preimage(hash: Hash) -> Option<Vec<u8>>;
+	/// Store the bytes of a preimage on chain.
+	fn note_preimage(bytes: Vec<u8>) -> Result<(), ()>;
+	/// Request that someone report a preimage.å
```
Comment (xlc) [inline frame/support/src/traits/misc.rs:581]:
```suggestion
	/// Request that someone report a preimage.
```

File: frame/preimage/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+	/// This preimage is managed by the system with a number of reference counters.
+	System(u32),
+}
+
+/// The preimage metadata.
+#[derive(Clone, Encode, Decode, TypeInfo, MaxEncodedLen)]
+pub struct Preimage<BoundedVec, AccountId, Balance> {
+	// The preimage we are storing.
+	pub preimage: BoundedVec,
+	// A reference counter for who depends on this resource.
```
Comment (kianenigma) [inline frame/preimage/src/lib.rs:63]:
```suggestion
	/// The preimage we are storing.
	pub preimage: BoundedVec,
	/// A reference counter for who depends on this resource.
```

and why is `BoundedVec` a generic?

File: frame/preimage/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+		///
+		/// If the preimage was previously requested, no fees or deposits are taken for providing
+		/// the preimage. Otherwise, a deposit is taken proportional to the size of the preimage.
+		#[pallet::weight(0)]
+		pub fn note_preimage(origin: OriginFor<T>, bytes: Vec<u8>) -> DispatchResultWithPostInfo {
+			// We accept a signed origin which will pay a deposit, or a root origin where a deposit
+			// is not taken.
+			let maybe_sender = Self::ensure_signed_or_manager(origin)?;
+			let bounded_vec =
+				BoundedVec::<u8, T::MaxSize>::try_from(bytes).map_err(|()| Error::<T>::TooLarge)?;
```
Comment (kianenigma) [inline frame/preimage/src/lib.rs:154]:
```suggestion
				BoundedVec::<u8, T::MaxSize>::try_from(bytes).map_err(|_| Error::<T>::TooLarge)?;
```

File: frame/preimage/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+		/// The user is not authorized to perform this action.
+		NotAuthorized,
+	}
+
+	/// The preimages stored by this pallet.
+	// TODO: Maybe store preimage metadata in its own storage.
+	#[pallet::storage]
+	pub(super) type Preimages<T: Config> = StorageMap<
+		_,
+		Identity, // TODO: Double Check
```
Comment (kianenigma) [inline frame/preimage/src/lib.rs:127]: should be fine if all keys are coming from `T::Hashing`.

File: frame/preimage/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+	BoundedVec,
+};
+use scale_info::TypeInfo;
+
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+
+pub use pallet::*;
+
+/// A type to note whether a preimage is owned by a user or the system.
```
Comment (kianenigma) [inline frame/preimage/src/lib.rs:49]: `Owner` might be a more intuitive name.

File: frame/preimage/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+	pub(super) type Preimages<T: Config> = StorageMap<
+		_,
+		Identity, // TODO: Double Check
+		T::Hash,
+		Preimage<BoundedVec<u8, T::MaxSize>, T::AccountId, BalanceOf<T>>,
+	>;
+
+	/// Any outstanding preimage requests.
+	#[pallet::storage]
+	pub(super) type Requests<T: Config> = StorageMap<
```
Comment (kianenigma) [inline frame/preimage/src/lib.rs:134]: seemingly requests can only be made by the manager, right?

Comment (kianenigma) [reply frame/preimage/src/lib.rs:134]: and what's the use case of the system requesting a pre-image multiple times?

Comment (shawntabrizi) [reply frame/preimage/src/lib.rs:134]: Multiple different pallets may want some preimage for different reasons. The API is designed to keep around the preimage until all reference counters on it are gone.

File: frame/preimage/src/lib.rs (hunk 3 above)
Comment (gavofyork) [reply frame/preimage/src/lib.rs:63]: I think `MaxSize` would be a better type param to be passing. `Preimage` struct should know that it's a `BoundedVec<u8>`, and only need to be told of the size.

File: frame/preimage/src/lib.rs (hunk 4 above)
Comment (gavofyork) [reply frame/preimage/src/lib.rs:154]: I prefer to be explicit over the type we are dismissing.

File: frame/preimage/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+		}
+	}
+
+	// Clear a preimage from the storage of the chain, returning any deposit that may be reserved.
+	//
+	// If `maybe_owner` is provided, we verify that it is the correct owner before clearing the
+	// data.
+	//
+	// If `maybe_owner` is not provided, this function cannot return an error.
+	fn do_clear_preimage(hash: T::Hash, maybe_owner: Option<T::AccountId>) -> DispatchResult {
```
Comment (gavofyork) [inline frame/preimage/src/lib.rs:316]: Should be named `maybe_check_owner`.

**Comment by gavofyork** [comment]
Just needs a few tests and I'd be happy to merge.

**Pull request closed without merging**